---
title: "Analysis of Drop-seq data"
output: html_document
---

In this document I will be detailing the analysis of Drop-seq data we received from Sisi Chen of the Thomson lab out of UCSF.  

#  Preliminary analysis

We have 4 datasets from 4 timepoints, multiplexed on an Illumina HiSeq.  The timepoints correspond to the following barcodes:

* Day 0: AAGGAGTA
* Day 1: ACTGCATA
* Day 5: CTAAGCCT
* Day 8: GTAAGGAG

Each timepoint was demultiplexed using Illumina software, resulting in 4 pairs of fastq files.  The first mate is 25 base pairs (bp) and the second mate is 50 bp.  The first mate should correspond to the 12 bp cell barcode and a 8 bp unique molecular identifier.  To verify the former, we can look at the distribution of cell barcodes for barcodes of length 10-14.  We can just look at one of the timepoints.

```{r eval=FALSE}
par(mfrow = c(1,5))
for(i in 10:14){
  x = paste("awk '{if (NR % 4 == 2) print $0}' AAGGAGTA_1.fastq | cut -c 1-", i, " | sort | uniq -c | awk -v OFS='\t' '{print $2, $1}' > AAGGAGTA_", i, "bp_barcode_counts.txt", sep = "")
  system(x)
  counts = read.table(file = paste("AAGGAGTA_", i, "bp_barcode_counts.txt", sep = ""))
  counts_hist = hist(counts[,2], breaks = 0:max(counts[,2]), xlim = c(0, 10))
}
```

### Head of counts histogram for first 10-14 bp of AAGGAGTA_1.fastq
```{r fig.width=6, echo=FALSE}
library(png)
library(grid)
img <- readPNG("AAGGAGTA_10-14bp_counts_hist_1-10.png")
 grid.raster(img)
```


The total number of counts looks as follows:

total counts | bp 
------------ | --
448106       | 10
721391       | 11
1094593      | 12
1940900      | 13
3745861      | 14

There seems to be a definite jump from 12 to13 bp, indicating that after 12bp the next base is more random than the previous base.  Later, when we take into account UMIs we'll investigate what minimum threshold is needed for a cell barcoded to be counted.

Let's take a look at the nucleotide and quality score distribution over all the timepoints using FastQC (http://www.bioinformatics.babraham.ac.uk/projects/fastqc/).

### Cell barcodes nucleotide distribution
```{r fig.width=4, echo=FALSE}
library(png)
library(grid)
img <- readPNG("barcodes_per_base_sequence_content.png")
 grid.raster(img)
```

### Cell barcodes quality distribution

```{r fig.width=4, echo=FALSE}
library(png)
library(grid)
img <- readPNG("barcodes_per_base_quality.png")
 grid.raster(img)
```

It seems like there are problems with the nucleotide content of the first few bases.  There is a vast preponderance of TAC in the first three bases (over $95 \%$ of the reads). This is likely due to biases in the experimental protocol of generating barcodes.  It shouldn't concern us too much, but the quality scores should.  The Drop-seq Alignment cookbook v1.1 (http://mccarrolllab.com/wp-content/uploads/2015/05/Drop-seqAlignmentCookbook_v1.1Aug2015.pdf) suggests dropping reads with any quality score below 10, corresponding to a 1 in 10 chance of the nucleotide being incorrect.  This may too lenient, we may want to restict ourselves to a higher quality score standard.  There are a total of 85,689,564 reads in the intial file.  If we restrict ourselves to cell barcodes that have quality scores all above 10 then we get 81,966,026 reads while if we restrict the quality scores to all above 20 then we get 51,876,289 reads. I think we can be liberal here and later remove cells with low barcode counts.  

Now let's look at the UMIs

### UMI nucleotide distribution
```{r fig.width=4, echo=FALSE}
library(png)
library(grid)
img <- readPNG("UMIs_per_base_sequence_content.png")
 grid.raster(img)
```

### UMI nucleotide quality distribution

```{r fig.width=4, echo=FALSE}
library(png)
library(grid)
img <- readPNG("UMIs_per_base_quality.png")
 grid.raster(img)
```

These look perfectly reasonable, though it looks like Ts get more common in later position and Gs get less common.

## Getting Digital Gene Expression (DGE)

We will first follow the recommendations of the Drop-seqAlignmentCookbook v1.1 (http://mccarrolllab.com/wp-content/uploads/2015/05/Drop-seqAlignmentCookbook_v1.1Aug2015.pdf).  Since there is an issue with the gtf format needed with software provided, we use the mm10 genome and gene reference available from McCarroll lab and associated with Gene Expression Omnibus accession GSE63472  (http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE63472).  We modify a few things from original pipeline.  We collapse cell barcodes and UMIs using an edit distance of 1.  We allow for more cells, or cell barcodes, and require a fewer number of distinct UMIs per cell barcode (10) to count a cell.


```{r engine='bash', eval=FALSE}
mkdir tmp

java -jar -Xmx8g ~/picard-tools-1.141/picard.jar FastqToSam F1=input_R1.fastq F2=input_R2.fastq SORT_ORDER=queryname O=input.sam SM=input TMP_DIR=tmp

rm -r tmp/

~/Drop-seq_tools-1.12/TagBamWithReadSequenceExtended INPUT=input.sam OUTPUT=input_barcoded.sam BASE_RANGE=1-12 BARCODED_READ=1 TAG_NAME=XB SUMMARY=input_cell_barcode_summary.txt BASE_QUALITY=10 NUM_BASES_BELOW_QUALITY=1

rm input.sam

~/Drop-seq_tools-1.12/TagBamWithReadSequenceExtended INPUT=input_barcoded.sam OUTPUT=input_barcoded_UMI.sam BASE_RANGE=13-20 BARCODED_READ=1 BASE_QUALITY=10 DISCARD_READ=TRUE TAG_NAME=XU

rm input_barcoded.sam

~/Drop-seq_tools-1.12/FilterBAM INPUT=input_barcoded_UMI.sam OUTPUT=input_filtered.sam

rm input_barcoded_UMI.sam

~/Drop-seq_tools-1.12/TrimStartingSequence I=input_filtered.sam O=input_filtered_ss_trimmed.sam SEQUENCE=AAGCAGTGGTATCAACGCAGAGTGAATGGG MISMATCHES=0 NUM_BASES=6
rm input_filtered.sam

~/Drop-seq_tools-1.12/PolyATrimmer I=input_filtered_ss_trimmed.sam O=input_trimmed.sam MISMATCHES=0 NUM_BASES=6

rm input_filtered_ss_trimmed.sam

java -Xmx8g -jar ~/picard-tools-1.141/picard.jar SamToFastq INPUT=input_trimmed.sam FASTQ=input_trimmed.fastq

mkdir mapped/

~/aligners/STAR/bin/Linux_x86_64_static/STAR --genomeDir ~/scratch/genomes/mm10/DropSeq/ --readFilesIn input_trimmed.fastq --outFileNamePrefix mapped/ --outSAMtype BAM Unsorted --sjdbGTFfile ~/scratch/genomes/mm10/DropSeq/mm10.gtf

java -jar -Xmx12g ~/picard-tools-1.141/picard.jar MergeBamAlignment UNMAPPED=input_trimmed.sam ALIGNED=mapped/Aligned.out.bam REFERENCE_SEQUENCE=~/scratch/genomes/mm10/DropSeq/mm10.fasta INCLUDE_SECONDARY_ALIGNMENTS=false PAIRED_RUN=false SORT_ORDER=queryname ALIGNED_READS_ONLY=true OUTPUT=input_mapped.bam

rm input_trimmed.sam

rm mapped/Aligned.out.bam

~/Drop-seq_tools-1.12/TagReadWithGeneExon I=input_mapped.bam O=input_mapped_gene_counts.bam TAG=GE ANNOTATIONS_FILE=~/scratch/genomes/mm10/DropSeq/mm10.gtf USE_STRAND_INFO=false ALLOW_MULTI_GENE_READS=false

rm input_mapped.bam

~/samtools-1.3/samtools sort input_mapped_gene_counts.bam -o input_mapped_gene_counts.sort.bam

rm input_mapped_gene_counts.bam

~/Drop-seq_tools-1.12/DetectBeadSynthesisErrors I=input_mapped_gene_counts.sort.bam O=input_mapped_clean_gene_counts.sort.bam SUMMARY=input_bead_clean_summary.txt CELL_BARCODE_TAG=XB MOLECULAR_BARCODE_TAG=XU GENE_EXON_TAG=GE MIN_UMIS_PER_CELL=10 NUM_BARCODES=10000 OUTPUT_STATS=input_bead_clean_out.txt

~/Drop-seq_tools-1.12/DigitalExpression SUMMARY=input_DGE_summary.txt O=input_DGE_matrix.txt I=input_mapped_clean_gene_counts.sort.bam CELL_BARCODE_TAG=XB MOLECULAR_BARCODE_TAG=XU GENE_EXON_TAG=GE EDIT_DISTANCE=1 NUM_CORE_BARCODES=10000
```

# Analysing the DGE

Now that we have digital gene expression profiles we can start analysing it.

```{r cache=TRUE, fig.height=4}
day0_DGE_matrix = read.table(file = "AAGGAGTA_DGE_matrix.txt", header = TRUE, row.names = 1)
day2_DGE_matrix = read.table(file = "ACTGCATA_DGE_matrix.txt", header = TRUE, row.names = 1)
day5_DGE_matrix = read.table(file = "CTAAGCCT_DGE_matrix.txt", header = TRUE, row.names = 1)
day8_DGE_matrix = read.table(file = "GTAAGGAG_DGE_matrix.txt", header = TRUE, row.names = 1)
# original sizes of the matrices
dim(day0_DGE_matrix)
dim(day2_DGE_matrix)
dim(day5_DGE_matrix)
dim(day8_DGE_matrix)

```

# Aggregate gene expression analysis

For a rough idea of the differences between time points we can combine the single cells into a single experiment and look at differential expression.

```{r cache=TRUE}
# treat combined single cells as a low-input RNA-seq experiment
day0_combined_DGE = rowSums(day0_DGE_matrix)
day2_combined_DGE = rowSums(day2_DGE_matrix)
day5_combined_DGE = rowSums(day5_DGE_matrix)
day8_combined_DGE = rowSums(day8_DGE_matrix)
# workaround to get all genes
all_days_combined_DGE = mat.or.vec(nr = length(union(names(day0_combined_DGE), union(names(day2_combined_DGE), union(names(day5_combined_DGE), names(day8_combined_DGE))))), nc = 4)
rownames(all_days_combined_DGE) = union(names(day0_combined_DGE), union(names(day2_combined_DGE), union(names(day5_combined_DGE), names(day8_combined_DGE))))
colnames(all_days_combined_DGE) = c("day0", "day2", "day5", "day8")

for(i in rownames(all_days_combined_DGE)){
  if(i %in% names(day0_combined_DGE)){
    all_days_combined_DGE[i, "day0"] = day0_combined_DGE[i]
  }}
for(i in rownames(all_days_combined_DGE)){
  if(i %in% names(day2_combined_DGE)){
    all_days_combined_DGE[i, "day2"] = day2_combined_DGE[i]
  }}
for(i in rownames(all_days_combined_DGE)){
  if(i %in% names(day5_combined_DGE)){
    all_days_combined_DGE[i, "day5"] = day5_combined_DGE[i]
  }}
for(i in rownames(all_days_combined_DGE)){
  if(i %in% names(day8_combined_DGE)){
    all_days_combined_DGE[i, "day8"] = day8_combined_DGE[i]
  }}

#filter missing genes
all_days_combined_DGE = all_days_combined_DGE[which(rowSums(all_days_combined_DGE) > 1), ]

# use DESeq2 to look for differentially expressed genes
source("https://bioconductor.org/biocLite.R")
biocLite("DESeq2")
library(DESeq2)
all_days_combined_coldata = data.frame("days" = c(0, 2, 5, 8))
rownames(all_days_combined_coldata) = c("day0", "day2", "day5", "day8")
all_days_combined_deseqdataseq = DESeqDataSetFromMatrix(countData=all_days_combined_DGE, colData= all_days_combined_coldata, design = ~ days)
# LRT, null model is empty, intercept only
all_days_combined_deseq = DESeq(all_days_combined_deseqdataseq, test="Wald")
all_days_combined_deseq_results = results(all_days_combined_deseq)
head(all_days_combined_deseq_results[order(all_days_combined_deseq_results$padj),])
all_days_combined_DGE[rownames(head(all_days_combined_deseq_results[order(all_days_combined_deseq_results$padj),])), ]
head(all_days_combined_deseq_results[order(all_days_combined_deseq_results$log2FoldChange, decreasing = TRUE),])
all_days_combined_DGE[rownames(head(all_days_combined_deseq_results[order(all_days_combined_deseq_results$log2FoldChange, decreasing = TRUE), ])), ]
```

```{r cache=TRUE, fig.width=5}
all_days_combined_top_genes = head(order(all_days_combined_deseq_results$padj), length(which(all_days_combined_deseq_results$padj < 0.05)))
all_days_combined_DGE[all_days_combined_top_genes, ]
# heatmap of top expressed genes
install.packages('pheatmap', repos = "http://cran.r-project.org")
library(pheatmap)
pheatmap(all_days_combined_DGE[all_days_combined_top_genes, ], cluster_col = FALSE)
```

There appears to be two major modules here, one in days 0, 2, and 5 consisting of Pim2, Igf2, Anxa2, Lgals1, Pou3f1, Krt18, Id1, and Nes and one in day 8 consisting of Bmp4, Slc30a2, Klf4, Alpl, and Calb2.  Additionally it appears to be two very specific genes, Gal for day 0 and Tdgf1 for day 8. 

```{r cache=TRUE, fig.width=5}
# increase FDR to 0.1
all_days_combined_top_genes = head(order(all_days_combined_deseq_results$padj), length(which(all_days_combined_deseq_results$padj < 0.1)))
# heatmap of top expressed genes
pheatmap(all_days_combined_DGE[all_days_combined_top_genes, ], cluster_col = FALSE)
```

# Single cell analysis

## How many cells to consider?

```{r cache=TRUE, fig.width=5}
# large number of low count cells
all_days_col_sum = c(colSums(day0_DGE_matrix), colSums(day2_DGE_matrix), colSums(day5_DGE_matrix), colSums(day8_DGE_matrix))
hist(all_days_col_sum, breaks = 200, ylim = c(0, 10000), xlab = "all days gene counts")
# how many counts to require to identify a cell?
# fit a negative binomial, set min count to be 90th percentile
library(MASS)
all_days_col_sum.nb_fit = fitdistr(all_days_col_sum, densfun = "negative binomial")
all_days_col_sum.thresh = qnbinom(0.99, size = all_days_col_sum.nb_fit$estimate[1], mu = all_days_col_sum.nb_fit$estimate[2])
day0_DGE_matrix = day0_DGE_matrix[ , colSums(day0_DGE_matrix) > all_days_col_sum.thresh]
day2_DGE_matrix = day2_DGE_matrix[ , colSums(day2_DGE_matrix) > all_days_col_sum.thresh]
day5_DGE_matrix = day5_DGE_matrix[ , colSums(day5_DGE_matrix) > all_days_col_sum.thresh]
day8_DGE_matrix = day8_DGE_matrix[ , colSums(day8_DGE_matrix) > all_days_col_sum.thresh]
dim(day0_DGE_matrix)
dim(day2_DGE_matrix)
dim(day5_DGE_matrix)
dim(day8_DGE_matrix)

# now we need to combine the days into a full dataset
length(intersect(intersect(intersect(rownames(day0_DGE_matrix), rownames(day2_DGE_matrix)), rownames(day5_DGE_matrix)), rownames(day8_DGE_matrix)))
length(union(union(union(rownames(day0_DGE_matrix), rownames(day2_DGE_matrix)), rownames(day5_DGE_matrix)), rownames(day8_DGE_matrix)))
# filter out genes that don't have many reads
day0_DGE_matrix = day0_DGE_matrix[rowSums(day0_DGE_matrix) > 10, ]
day2_DGE_matrix = day2_DGE_matrix[rowSums(day2_DGE_matrix) > 10, ]
day5_DGE_matrix = day5_DGE_matrix[rowSums(day5_DGE_matrix) > 10, ]
day8_DGE_matrix = day8_DGE_matrix[rowSums(day8_DGE_matrix) > 10, ]
names(day0_DGE_matrix) = paste0("day0_", names(day0_DGE_matrix))
names(day2_DGE_matrix) = paste0("day2_", names(day2_DGE_matrix))
names(day5_DGE_matrix) = paste0("day5_", names(day5_DGE_matrix))
names(day8_DGE_matrix) = paste0("day8_", names(day8_DGE_matrix))
```

# Clustering cells

## Principal compenent analysis (PCA)
```{r fig.width=5, cache=TRUE}
# PCA
# log10 scale first
# Note that we need to use R/3.0.2, get error "Error in La.svd(x, nu, nv) : error code 1 from Lapack routine 'dgesdd'" with version 3.2.2
day0_log10_norm_GE = log10(as.matrix(day0_DGE_matrix[rowSums(day0_DGE_matrix > 0) > 5, ]) + 1)
day0_log10_norm_GE_pca = prcomp(t(day0_log10_norm_GE), corr = TRUE, scale = TRUE)
plot(day0_log10_norm_GE_pca$x[,1], day0_log10_norm_GE_pca$x[,2], pch = 16, main = "day 0 PCA")
```

There doesn't appear to be any groupings here.  Let's look at the effect of sampling depth (number of UMIs per cell) has on the PCA.

```{r fig.width=5, cache=TRUE}
day0_DGE_lib_size = colSums(day0_DGE_matrix)
day0_DGE_lib_size = day0_DGE_lib_size - min(day0_DGE_lib_size) + 1
day0_DGE_lib_size = day0_DGE_lib_size/max(day0_DGE_lib_size)
day0_DGE_lib_size = round(day0_DGE_lib_size*100)
plot(day0_log10_norm_GE_pca$x[,1], day0_log10_norm_GE_pca$x[,2], pch = 16, main = "day 0 PCA colored by sampling depth", col = day0_DGE_lib_size)
```


It looks like the first principal component is directly accounting for the sampling depth.  In fact, in all time points the correlation between the two is nearly perfectly -1.

```{r cache=TRUE}
day2_log10_norm_GE = log10(as.matrix(day2_DGE_matrix[rowSums(day2_DGE_matrix > 0) > 5, ]) + 1)
day2_log10_norm_GE_pca = prcomp(t(day2_log10_norm_GE), corr = TRUE, scale = TRUE)
day5_log10_norm_GE = log10(as.matrix(day5_DGE_matrix[rowSums(day5_DGE_matrix > 0) > 5, ]) + 1)
day5_log10_norm_GE_pca = prcomp(t(day5_log10_norm_GE), corr = TRUE, scale = TRUE)
day8_log10_norm_GE = log10(as.matrix(day8_DGE_matrix[rowSums(day8_DGE_matrix > 0) > 5, ]) + 1)
day8_log10_norm_GE_pca = prcomp(t(day8_log10_norm_GE), corr = TRUE, scale = TRUE)

c(cor(day0_log10_norm_GE_pca$x[,1], colSums(day0_DGE_matrix)), cor(day2_log10_norm_GE_pca$x[,1], colSums(day2_DGE_matrix)), cor(day5_log10_norm_GE_pca$x[,1], colSums(day5_DGE_matrix)), cor(day8_log10_norm_GE_pca$x[,1], colSums(day8_DGE_matrix)) )
```

Since the first factor is effectively the sampling depth we can look at the second and third components.

```{r fig.width=5, cache=TRUE}
plot(day0_log10_norm_GE_pca$x[,2], day0_log10_norm_GE_pca$x[,3], pch = 16, main = "day 0 PCA 2nd & 3rd")
```

There doesn't appear to be definitive clusters of cells, except for the possibility that there is one major cluster with outliers.  It may be that we need to properly normalize the gene expression counts to account for sampling depth.

